support/on-prem-archive.sh

```diff
@@ -0,0 +1,122 @@
+#!/bin/bash
+
+# The purpose of this script is to download the latest stable version of every
+# package in the core-plans repo, tar them up, and upload to S3. It also supports
+# downloading the archive from S3, extracting it, and uploading to a new depot.
+
+set -eo pipefail
+
+help() {
+    echo "Usage: on-prem-archive.sh {create-archive|populate-depot <DEPOT_URL>}"
+}
+
+exists() {
+  if command -v $1 >/dev/null 2>&1
+  then
+    return 0
+  else
+    return 1
+  fi
+}
+
+s3_cp() {
+  aws s3 cp --acl=public-read ${1} ${2} >&2
+}
+
+check_tools() {
+  for tool in "$1"
+  do
+    if ! exists "$tool"; then
+      echo "Please install $tool and run this script again."
+      exit 1
+    fi
+  done
+}
+
+check_vars() {
+  for var in "$1"
+  do
+    if [ -z "${!var}" ]; then
+      echo "Please ensure that $var is exported in your environment and run this script again."
+      exit 1
+    fi
+  done
+}
+
+bucket="habitat-on-prem-builder-bootstrap"
+marker="LATEST.tar.gz"
+
+case "$1" in
+  create-archive)
+    required_tools=( aws git curl jq )
+    required_vars=( AWS_ACCESS_KEY_ID AWS_SECRET_ACCESS_KEY )
+    core_tmp=$(mktemp -d)
+    core="$core_tmp/core-plans"
+    upstream_depot="https://bldr.habitat.sh"
+    bootstrap_file="on-prem-bootstrap-$(date +%Y%m%d%H%M%S).tar.gz"
+    tar_file="/tmp/$bootstrap_file"
+    tmp_dir=$(mktemp -d)
+
+    check_tools ${required_tools[@]}
+    check_vars ${required_vars[@]}
+
+    git clone https://github.com/habitat-sh/core-plans.git "$core"
+    pushd "$core"
+    for p in $(find . -maxdepth 1 -type d -exec basename '{}' \; | sort)
+    do
+      latest=$(curl -H "Accept: application/json" "$upstream_depot/v1/depot/channels/core/stable/pkgs/$p/latest")
+      raw_ident=$(echo "$latest" | jq ".ident")
+
+      if [ "$raw_ident" = "null" ]; then
+        continue
+      fi
+
+      slash_ident=$(echo "$raw_ident" | jq "\"\(.origin)/\(.name)/\(.version)/\(.release)\"" | tr -d '"')
+      dash_ident=$(echo "$raw_ident" | jq "\"\(.origin)-\(.name)-\(.version)-\(.release)\"" | tr -d '"')
+      target=$(echo "$latest" | jq ".target" | tr -d '"')
+
+      curl -H "Accept: application/json" -o $tmp_dir/$dash_ident-$target.hart "$upstream_depot/v1/depot/pkgs/$slash_ident/download"
+    done
+    popd
+
+    cd /tmp
+    tar zcvf $tar_file -C $tmp_dir .
+    s3_cp $tar_file s3://$bucket/
+    s3_cp s3://$bucket/$bootstrap_file s3://$bucket/$marker
+    rm -fr "$tmp_dir"
```

Comment:
Since we'll have waited a long time for all of these to download, should we allow for keeping these around? (I think the thing I'm running right now is going to fail because of a name conflict, so if it does, it'd be good not to have to download everything again.)

Reply:
Yeah, I just ran this and got ```upload failed: ../../tmp/on-prem-bootstrap-20180212121935.tar.gz to s3://habitat-on-prem-builder-bootstrap/on-prem-bootstrap-20180212121935.tar.gz An error occurred (AccessDenied) when calling the CreateMultipartUpload operation: Access Denied``` I had commented the removals, though, so luckily I still have the 4 GB tar file.

Reply:
Sure

Reply:
`HAB_ON_PREM_BOOTSTRAP_DONT_CLEAN_UP` will skip the removal of things now.